quic/s2n-quic/src/tests/dc.rs

```diff
@@ -0,0 +1,169 @@
+    let server_tokens = [TEST_TOKEN_1];
+    let client_tokens = [TEST_TOKEN_2];
+
+    test(model, |handle| {
+        let server_tls = build_server_mtls_provider(certificates::MTLS_CA_CERT)?;
+        let mut server = Server::builder()
+            .with_io(handle.builder().build()?)?
+            .with_tls(server_tls)?
+            .with_event((tracing_events(), server_subscriber))?
+            .with_random(Random::with_seed(456))?
+            .with_dc(MockDcEndpoint::new(&server_tokens))?
+            .start()?;
+
+        let addr = server.local_addr()?;
+        spawn(async move {
+            let mut conn = server.accept().await.unwrap();
+            let mut stream = conn.open_bidirectional_stream().await.unwrap();
+            stream.send(vec![42; LEN].into()).await.unwrap();
+            stream.flush().await.unwrap();
+        });
+
+        let client_tls = build_client_mtls_provider(certificates::MTLS_CA_CERT)?;
+        let client = Client::builder()
+            .with_io(handle.builder().build().unwrap())?
+            .with_tls(client_tls)?
+            .with_event((tracing_events(), client_subscriber))?
+            .with_random(Random::with_seed(456))?
+            .with_dc(MockDcEndpoint::new(&client_tokens))?
+            .start()?;
+
+        primary::spawn(async move {
+            let connect = Connect::new(addr).with_server_name("localhost");
+            let mut conn = client.connect(connect).await.unwrap();
+            let mut stream = conn.accept_bidirectional_stream().await.unwrap().unwrap();
+
+            let mut recv_len = 0;
+            while let Some(chunk) = stream.receive().await.unwrap() {
+                recv_len += chunk.len();
+            }
+            assert_eq!(LEN, recv_len);
+        });
+
+        Ok(addr)
+    })
+    .unwrap();
+
+    let server_events = server_events.events().lock().unwrap().clone();
+    let client_events = client_events.events().lock().unwrap().clone();
+
+    // 3 state transitions (VersionNegotiated -> PathSecretsReady -> Complete)
+    assert_eq!(3, server_events.len());
+    assert_eq!(3, client_events.len());
+
+    for events in [server_events.clone(), client_events.clone()] {
+        if let DcState::VersionNegotiated { version, .. } = events[0].state {
+            assert_eq!(version, s2n_quic_core::dc::SUPPORTED_VERSIONS[0]);
+        } else {
+            panic!("VersionNegotiated should be the first dc state");
+        }
+
+        assert!(matches!(events[1].state, DcState::PathSecretsReady { .. }));
+        assert!(matches!(events[2].state, DcState::Complete { .. }));
+    }
+
+    // Server completes in 2.5 RTTs measured from the start of the test, since it takes .5 RTT
+    // for the Initial from the client to reach the server
+    assert_eq!(
+        rtt.mul_f32(2.5),
+        server_events[2].timestamp.duration_since_start()
+    );
+    assert_eq!(rtt * 2, client_events[2].timestamp.duration_since_start());
```

Comment:
Very nice assertions :) Can we also assert when the path secrets are ready as well?